src/map.cpp

```diff
@@ -529,62 +529,138 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
-
-	const int8_t mx = start.x < destination.x ? 1 : start.x == destination.x ? 0 : -1;
-	const int8_t my = start.y < destination.y ? 1 : start.y == destination.y ? 0 : -1;
+	if (blockFloor && tile->getGround()) {
+		return false;
+	}
 
-	int32_t A = Position::getOffsetY(destination, start);
-	int32_t B = Position::getOffsetX(start, destination);
-	int32_t C = -(A * destination.x + B * destination.y);
+	return !tile->hasProperty(CONST_PROP_BLOCKPROJECTILE);
+}
 
-	while (start.x != destination.x || start.y != destination.y) {
-		int32_t move_hor = std::abs(A * (start.x + mx) + B * (start.y) + C);
-		int32_t move_ver = std::abs(A * (start.x) + B * (start.y + my) + C);
-		int32_t move_cross = std::abs(A * (start.x + mx) + B * (start.y + my) + C);
+bool Map::getSteepLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		if (start.y != destination.y && (start.x == destination.x || move_hor > move_ver || move_hor > move_cross)) {
-			start.y += my;
+	for (uint16_t y = x0 + 1; y < x1; ++y) {
+		if (!isTileClear(std::floor(xy), y, z)) {
+			return false;
 		}
+		xy += grad;
+	}
 
-		if (start.x != destination.x && (start.y == destination.y || move_ver > move_hor || move_ver > move_cross)) {
-			start.x += mx;
-		}
+	return true;
+}
+
+bool Map::getSlightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
+	for (uint16_t x = x0 + 1; x < x1; ++x) {
+		if (!isTileClear(x, std::floor(xy + 0.1), z)) {
 			return false;
 		}
+		xy += grad;
 	}
 
-	// now we need to perform a jump between floors to see if everything is clear (literally)
-	while (start.z != destination.z) {
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->getThingCount() > 0) {
-			return false;
+	return true;
+}
+
+bool Map::checkSightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	if (x0 == x1 && y0 == y1) {
+		return true;
+	}
+
+	if (std::abs(y1 - y0) > std::abs(x1 - x0)) {
+		if (y1 > y0) {
+			return getSteepLine(y0, x0, y1, x1, z);
 		}
+		return getSteepLine(y1, x1, y0, x0, z);
+	}
 
-		start.z++;
+	if (x0 > x1) {
+		return getSlightLine(x1, y1, x0, y0, z);
 	}
 
-	return true;
+	return getSlightLine(x0, y0, x1, y1, z);
 }
 
-bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck) const
+bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck /*= false*/) const
 {
-	if (floorCheck && fromPos.z != toPos.z) {
+	uint16_t x0 = fromPos.x;
+	uint16_t y0 = fromPos.y;
+	uint8_t  z0 = fromPos.z;
+
+	uint16_t x1 = toPos.x;
+	uint16_t y1 = toPos.y;
+	uint8_t  z1 = toPos.z;
```

Comment:
Just use `fromPos.x`, `toPos.y` etc., the only aux values you need is `z0` as you modify it